tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,176 @@
+            module_path: package_dir,
+            sdk_image: sdk,
+            go_mod_cache: &root_dir.join(".gomodcache"),
+            command: format!(
+                "tar xf {input} &&
+                pushd {moddir} &&
+                export GOPROXY={goproxy} &&
+                export GOSUMDB={gosumdb} &&
+                export GOPRIVATE={goprivate} &&
+                go list -mod=readonly ./... >/dev/null && go mod vendor &&
+                popd &&
+                tar czf {output} {moddir} && 
+                rm -rf {moddir}",
+                input = path_arg.to_string_lossy(),
+                moddir = mod_dir.to_string_lossy(),
+                goproxy = goproxy,
+                gosumdb = gosumdb,
+                goprivate = goprivate,
+                output = output_path_arg.to_string_lossy(),
+            ),
+        };
+        docker_go(root_dir, &args)?;
+
+        Ok(())
+    }
+}
+
+fn extract_file_name(url: &str) -> Result<PathBuf> {
+    let parsed = reqwest::Url::parse(url).context(error::InputUrlSnafu { url })?;
+    let name = parsed
+        .path_segments()
+        .context(error::InputFileBadSnafu { path: url })?
+        .last()
+        .context(error::InputFileBadSnafu { path: url })?;
+    Ok(name.into())
+}
+
+struct DockerGoArgs<'a> {
+    module_path: &'a Path,
+    sdk_image: String,
+    go_mod_cache: &'a Path,
+    command: String,
+}
+
+/// Run `docker-go` with the specified arguments.
+fn docker_go(root_dir: &Path, dg_args: &DockerGoArgs) -> Result<Output> {
+    let args = vec![
+        "--module-path",
+        dg_args
+            .module_path
+            .to_str()
+            .context(error::InputFileSnafu)?,
+        "--sdk-image",
+        &dg_args.sdk_image,
+        "--go-mod-cache",
+        dg_args
+            .go_mod_cache
+            .to_str()
+            .context(error::InputFileSnafu)?,
+        "--command",
+        &dg_args.command,
+    ];
+    let arg_string = args.join(" ");
+    let program = root_dir.join("tools/docker-go");
+    println!("program: {}", program.to_string_lossy());
+    let output = cmd(program, args)
+        .stderr_to_stdout()
+        .stdout_capture()
+        .unchecked()
+        .run()
+        .context(error::CommandStartSnafu)?;
+
+    let stdout = String::from_utf8_lossy(&output.stdout);
+    println!("{}", &stdout);
+    ensure!(
+        output.status.success(),
+        error::DockerExecutionSnafu { args: arg_string }
+    );
+    Ok(output)
+}
+
+/// Run `go env` with the specified argument.
+fn go_env(var: &str) -> Option<String> {
+    let args = vec!["env", var];
+    let output = match cmd("go", args)
+        .stderr_to_stdout()
+        .stdout_capture()
+        .unchecked()
+        .run()
+    {
+        Ok(v) => v,
+        Err(_) => return None,
+    };
+    let stdout = String::from_utf8_lossy(&output.stdout);
+    println!("{}", &stdout);
+    output.status.success().then(|| stdout.to_string())
+}
```

Comment:
We don't want to assume that a `go` install exists on the host or works as expected.

Reply:
Gotcha - what would good indicators that buildsys shouldn't proceed with vendoring the go module on the host?

My first thought is to check:
- is the `go` tool in the path?
- is the output of `go version` greater than 1.16?

If either of the answer to those is "no", we fail and exit early.

Reply:
Hmm, is this redundant with the checks in `docker-go`? That script already has the logic to use `go` if we have it, otherwise read the environment variable directly.

Reply:
👍🏼 Removed this block since `docker-go` is doing the same thing and gets the necessary go env vars